src/librustc_mir/transform/validate.rs

```diff
@@ -99,20 +217,37 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
     }
 
     fn visit_statement(&mut self, statement: &Statement<'tcx>, location: Location) {
-        // The sides of an assignment must not alias. Currently this just checks whether the places
-        // are identical.
-        if let StatementKind::Assign(box (dest, rvalue)) = &statement.kind {
-            match rvalue {
-                Rvalue::Use(Operand::Copy(src) | Operand::Move(src)) => {
-                    if dest == src {
-                        self.fail(
-                            location,
-                            "encountered `Assign` statement with overlapping memory",
-                        );
+        match &statement.kind {
+            StatementKind::Assign(box (dest, rvalue)) => {
```

Comment:
As a follow up we should also check function calls and `DropAndReplace`

Reply:
`SetDiscriminant` can also get checked as to whether the type has the mentioned discriminant id and whether the local's type is actually an enum

Reply:
Let's collect these ideas in https://github.com/rust-lang/rust/issues/73832